p521/src/arithmetic/field.rs

```diff
@@ -233,7 +233,22 @@ impl FieldElement {
     /// Returns the square root of self mod p, or `None` if no square root
     /// exists.
     pub fn sqrt(&self) -> CtOption<Self> {
-        todo!("`sqrt` not yet implemented")
+        // Tonelli-Shank's algorithm for q mod 4 = 3 (i.e. Shank's algorithm)
+        // https://eprint.iacr.org/2012/685.pdf
+        let w = self.pow_vartime(&[
+            0x00000000_00000000,
+            0x00000000_00000000,
+            0x00000000_00000000,
+            0x00000000_00000000,
+            0x00000000_00000000,
+            0x00000000_00000000,
+            0x00000000_00000000,
+            0x00000000_00000000,
+            0x00000000_00000040,
```

Comment:
Needs to end with `80` because it's 2^519.

Reply:
Hmm, interesting, substituting:

```suggestion
            0x00000000_00000080,
```

(not sure I follow why, but)

...passes sqrt for 1 and 4 but fails on 9:

```
---- arithmetic::field::tests::sqrt stdout ----
thread 'arithmetic::field::tests::sqrt' panicked at 'assertion failed: `(left == right)`
  left: `FieldElement([8, 0, 288230376151711744, 288230376151711743, 288230376151711743, 288230376151711743, 288230376151711743, 288230376151711743, 144115188075855871])`,
 right: `FieldElement([9, 0, 0, 0, 0, 0, 0, 0, 0])`', p521/src/arithmetic/field.rs:574:5
```

Reply:
9 fails due to https://github.com/RustCrypto/elliptic-curves/pull/946#issuecomment-1789836157

This will fix it

Reply:
Well, indeed it all works, the sqrt tests are passing, although I'm still not sure why it's (p+1)/4 instead of (q+1)/4

Reply:
Assuming our number is a quadratic residue => `x^((p-1)/2)) = 1` due to Euler's criterion. Therefore `x^((p+1)/2)) = x`.

There's a simplification that can be made when `p=3mod4` in which case p can be expressed as `4k +3` which guarantees  `(p+1)/4` to be an integer in which case `(x^((p+1)/4))^2 = x`. This means `x^((p+1)/4)` is the square root.